classify/blast.py

```diff
@@ -38,44 +46,73 @@ class BlastnTool(BlastTools):
     """ Tool wrapper for blastn """
     subtool_name = 'blastn'
 
-    def get_hits_pipe(self, inPipe, db, threads=None):
-
+    def get_hits_pipe(self, inPipe, db, threads=None, task=None, outfmt='6', max_target_seqs=1, output_type="read_id"):
+        _log.debug("Executing get_hits_pipe function.")
+        #Validates output type run during this function
+        _log.debug(f"get_hits_pipe called with outfmt: {outfmt}")
+        #toggle between extracting read IDs only or full blast output (all lines)
+        if output_type not in ['read_id', 'full_line']:
+            _log.warning(f"Invalid output_type '{output_type}' specified. Defaulting to 'read_id'.")
+            output_type = 'read_id'
         # run blastn and emit list of read IDs
         threads = util.misc.sanitize_thread_count(threads)
         cmd = [self.install_and_get_path(),
             '-db', db,
             '-word_size', 16,
             '-num_threads', threads,
             '-evalue', '1e-6',
-            '-outfmt', 6,
-            '-max_target_seqs', 1,
+            '-outfmt', str(outfmt),
+            '-max_target_seqs', str(max_target_seqs),
+            '-task', str(task) if task else 'blastn',
         ]
         cmd = [str(x) for x in cmd]
+        #Log BLAST command executed
+        _log.debug('Running blastn command: {}'.format(' '.join(cmd)))
         _log.debug('| ' + ' '.join(cmd) + ' |')
         blast_pipe = subprocess.Popen(cmd, stdin=inPipe, stdout=subprocess.PIPE)
+        output, error = blast_pipe.communicate()
 
-        # strip tab output to just query read ID names and emit
-        last_read_id = None
-        for line in blast_pipe.stdout:
-            line = line.decode('UTF-8').rstrip('\n\r')
-            read_id = line.split('\t')[0]
-            # only emit if it is not a duplicate of the previous read ID
-            if read_id != last_read_id:
-                last_read_id = read_id
-                yield read_id
-
-        if blast_pipe.poll():
+        #Display error message if BLAST failed
+        if blast_pipe.returncode!= 0:
+            _log.error('Error running blastn command: {}'.format(error))
             raise subprocess.CalledProcessError(blast_pipe.returncode, cmd)
-
+        
+        # If read_id is defined, strip tab output to just query read ID names and emit (default)
+        if output_type == 'read_id':
+            last_read_id = None
+            for line in output.decode('UTF-8').splitlines():
+                line = line.strip()
```

Comment:
Yes, but line.strip() also removes any trailing whitespace, including tabs, so this might still be useful to keep.

Reply:
I don't think we'd want that though, in part because those are meaningful characters, and also in part because it doesn't really matter anyway if we follow it up with `line.split('\t')[0]`

Reply:
Ok, sounds good, removed line.strip()